bokeh/server/views/autoload_js_handler.py

```diff
@@ -70,14 +76,17 @@ class AutoloadJsHandler(SessionHandler):
         resources = self.application.resources(server_url) if resources_param != "none" else None
         bundle = bundle_for_objs_and_resources(None, resources)
 
-        render_items = [RenderItem(sessionid=session.id, elementid=element_id, use_for_title=False)]
+        render_items = [RenderItem(token=session.token, elementid=element_id, use_for_title=False)]
```

Comment:
@bryevdv @philippjfr I really don't think it's anywhere near close to being secure. The token ends up embedded directly in the HTML code. It means that *any* browser extension that can read data on the page can now get to that token and (with additional permissions, I think) send requests on behalf of the web page (or just embed JS that would send such requests). What's even worse - any JS already embedded within the page can now get to the token as well, and even more so - it definitely can send requests using it. Even Facebook tracker could access it.

I have mentioned it before - a proper solution is to use cookies that are `HttpOnly` and `Secure`. The latter requires HTTPS that's usually not available on `localhost` though. But we really need to push HTTPS for anything that's not on `localhost` - without it, auth just doesn't make sense.
Using cookies is easy with `bokeh serve`. It's a bit more involved when embedding Bokeh. It seems likely that a user will have to ask Bokeh for a cookie that will be then correctly applied by the user.

Reply:
@p-himik I am willing to believe that  there are improvements that can be made, however I'd like to compare the current situation to the previous situation and see if you agree that this is an improvement, as it stands. 

* Previously, a perpetual, non-expirable token was present in URL query arguments, and thus in any number of logging systems that might be in between the client and server. 

* Currently, the token can now be set to expire (default: 5 minutes, but could be set to shorter) after which it is useless and cannot be used to create sessions, and does not appear in any URL query arguments (will not spill in to logs)

> But we really need to push HTTPS for anything that's not on localhost - without it, auth just doesn't make sense.

I don't disagree, but many people do not care about auth at all, and just want to publish something open on the internet. It's much easier for them to get HTTPS by just standing up a vanilla server behind a proxy that can terminate SSL. If we say everyone who runs a Bokeh server for any reason has to install SSL certs we have raised the bar considerably.

Reply:
I agree with your points about this not being secure and at least for me that was not the primary concern in this PR (accessing cookies and headers and adding other data was). So I'll join @bryevdv in asking whether you're saying that we actively made the situation worse with this PR, i.e. wouldn't any extension have been able to access the token from the URL parameter previously?

Reply:
And if not, are there any quick high value,  low-risk  changes you would suggest for the immediate term (e.g. lowering the default expiry, etc)

Reply:
> Previously, a perpetual, non-expirable token was present in URL query arguments, and thus in any number of logging systems that might be in between the client and server.

If using HTTP instead of HTTPS, same logging systems might intercept the current token since it would be in plain text in the HTTP GET body.
IIRC query parameters and even the path are encrypted when using HTTPS.
So this particular item is not really an improvement. And in any case, it could be alleviated by using `--allow-websocket-origin`, which should be used anyway, even without auth considerations.

> Currently, the token can now be set to expire

True. But once you open a WS connection, the token is not checked anymore. If any malicious script happens to be able to access the webpage, it would be able to immediately create a WS connection and do whatever until the user interrupts it (by closing the tab, most likely).

> many people do not care about auth at all, and just want to publish something open on the internet

For people that don't use auth, pushing HTTPS is unnecessary, I agree. Although avoiding plain HTTP is still useful here, but that's out of the scope of the current issue and even Bokeh.

> standing up a vanilla server behind a proxy that can terminate SSL

I admit, I have barely any experience here, so cannot really tell anything. I think Bokeh should be able to check that a secure endpoint is used by checking the `Origin` header, but I'm not 100% sure. Maybe there are other ways to make sure that only a secure endpoint is used. If there are no such ways, there can be a flag like `--allow-auth-over-http` that's checked only when the auth functionality is used.

> are you proposing that we should not release on Monday?

I propose that the auth functionality should at least be made alpha and documented as such, with possible vulnerabilities mentioned above outlined in the documentation.
Otherwise, it's a false promise of security.

Reply:
@philippjfr 
> asking whether you're saying that we actively made the situation worse with this PR, i.e. wouldn't any extension have been able to access the token from the URL parameter previously?

You haven't made it worse, as far as I can tell. You just did not improve it in any way for the end users.

@bryevdv 
> are there any quick high value, low-risk changes you would suggest for the immediate term

Documenting, admitting the alpha and possible insecurities, ideally issuing a warning somewhere where even the most documentation-avoiding users would see it.

Reply:
> You haven't made it worse, as far as I can tell. You just did not improve it in any way for the end users.

Okay good to know, I'd hope in the long term we can then add a configuration option which switches to using cookies to authorize the session creation. The token will then only contain the signed session information when this option is disabled, but still be useful for exposing headers and cookies to the enduser.

Reply:
Maybe our workflows with Bokeh differ dramatically, but I have no idea why you imply that cookies cannot be used all the time, without any optional flags.

> The token will then only contain the signed session information

With cookies, it is not necessary. Modern browsers are quite good at preventing anyone from stealing cookies. Much better than they are at preventing someone stealing a plain text token in raw HTML.

> also still be useful for exposing headers and cookies to the enduser

The end user does not need authentication cookies on the frontend. On the backend, they only need to provide the cookies to Bokeh, and even that is need only if bokeh is embedded within some other server, like Flask or Django.
Regarding exposing headers - I have no idea what you mean. You can already access headers in Python because a document instance has access to the initial request object.

Reply:
At least when you're running multiple processes (using `--num-procs`) this is not the case. The header and cookies will not necessarily be available to the process the Websocket request arrives on because it may be a different process the initial HTTP request arrived on.

Reply:
> Of using HTTP instead of HTTPS, same logging systems might intercept the current token since it would be in plain text in the HTTP GET body.

Just to note, *lots* of systems use bearer tokens in headers, and what you say is true for them as well. I am not concerned about any lack of security in the HTTP case, lack of security is *given* in the HTTP case. 

I think we should be clear about the terminology and separation of concerns. There are potentially different usage of "auth".

* One is the "auth hooks"&mdash;these just expose the exact hooks already well-established in tornado to Bokeh users. These are generic and can be used to enforce any auth flow purely on the HTTP(s) side of things, before any websocket connection is ever tried. The onus is on users to implement the auth flow they need, but I don't think there is *anything* alpha about this part.

* The other (this) is for a much narrower case, only for passing a bearer token that authorize a websocket connection, after any up-front auth flow has succeeded. This, specifically, is where your concerns lie. 

 Can we agree on this, first of all?

---

Then I'd like to characterize the risk so it can be communicated.  Coming back to 
>any JS already embedded within the page can now get to the token as well, and even more so

Is this actually a concern? What other JS would be on the page, except that which author of the page, i.e the person embedding the Bokeh content into that page, decided to put there, themself? If a person (or organization) wants to own themself, that's their business. 

So that leaves the browser extension possibility. And I agree that is reasonable to outline. how about this disclaimer:

> When embedding Bokeh server content in to pages, an expirable webscoekt session token is included in the rendered page. This token could be accessed potentially be accessed by browser extensions with sufficient privileges. It is recommended to always use SSL termination with an auth flow on the front-end, and to set the websocket  token expiry as short as possible. Future work on the Bokeh server will afford the option of transmitting the token in secure cookies (see issue #xxxx)

Thoughts?

Reply:
I do disagree with this: 

> You just did not improve it in any way for the end users.

Going from an easily spillable, un-expirable token to an exportable token that requires someone to write and install a browser extension to make use of it (if they can, before it expires) is an improvement. Security is always about a frontier of risk. I want to be honest and up front about any risks but also specific (including about where and how that frontier may have shrunk)

Reply:
Or to characterize things another way, here is my understanding:

If these things are true:
* auth hooks are set up
* SSL termination is enabled
* allow websocket origin is narrowly specified 
* you control the page content and domain that embeds the bokeh app

Then the only real avenue is for *your trusted user* to intentionally have installed a browser plugin to connect to the exact same content they are already looking at in their authorized session, and only for as long as the tab stays open (assuming it could be used for expiry to begin with). 

Asking (sincerely) to poke holes in that scenario.

Reply:
> Can we agree on this, first of all?

Yes, I only mean the second, narrower, case.


But after thinking about it some more, I realized that with embedded malicious code secure cookies help just somewhat - they don't really solve the issue since the scenarios that I've described are basically XSS attacks that already happened and were successful.

In that case, you're probably right.
I really hope that I'm just in the deepest Dunning–Kruger valley right now and not doubting myself unjustifiably. And if that's indeed the case, I do apologize for all the ruckus.

I guess nothing needs to be done right now in that case.
I'll try to find a security expert that would agree to check this functionality. No promises though. :)

Reply:
Your input here is hugely appreciated, I'm certainly no expert in this area and primarily tackled this PR for reasons other than security. If we could get a security expert to chime in that would be incredibly helpful.

Reply:
> I really hope that I'm just in the deepest Dunning–Kruger valley right now and not doubting myself unjustifiably. And if that's indeed the case, I do apologize for all the ruckus. 

Absolutely *no apology* necessary. This is a topic that only benefits from being aired out and being thinked on by different people and from different perspectives. For the same of complete transparency I will add a note similar to the above somewhere in the server user guide docs so that it can be referred to and users can make their own risk assessments. 

> I'll try to find a security expert that would agree to check this functionality. No promises though. :)

That would be amazing, welcome, and incredibly appreciated. I think we could even discuss allocating some project funds as an inducement, if that helps.